src/main/scala/fi/vm/sade/hakurekisteri/integration/koski/KoskiService.scala

```diff
@@ -350,6 +357,48 @@ class KoskiService(
         logger.error("Kutsu koskeen oppijanumeroille {} epäonnistui: {} ", oppijaOids, e)
         Future.failed(e)
       }
+  }
+
+  def resolveKoulusivistyskieli(henkilo: KoskiHenkiloContainer): Seq[String] = {
+    val validitSuoritukset = henkilo.opiskeluoikeudet.flatMap(o =>
+      o.suoritukset.filter(s => s.isLukionOrPerusopetuksenoppimaara())
+    )
+
+    validitSuoritukset.flatMap(s => s.koulusivistyskieli.map(k => k.map(_.koodiarvo))).flatten
+  }
+
+  private def fetchKoulusivistyskieletForReal(
+    oppijaOids: Seq[String]
+  ): Future[Map[String, Seq[String]]] = {
+    fetchHenkilot(oppijaOids.toSet).map(
+      _.map(h => (h.henkilö.oid.get, resolveKoulusivistyskieli(h))).toMap
+    )
+  }
+
+  override def fetchKoulusivistyskielet(
+    oppijaOids: Seq[String]
+  ): Map[String, Seq[String]] = {
+    logger.info(s"Pyydetty Koskesta koulusivistyskieli ${oppijaOids.size} henkilölle.")
+    val cached = koskiKoulusivistyskieliCache.getAllPresent(oppijaOids)
+    val missing = oppijaOids.filterNot(cached.keys.toSet.contains(_))
+    logger.info(
+      s"Välimuistista saatu ${cached.keys.size} henkilön koulusivistyskieli, haetaan suoraan Koskesta tiedot ${missing.size} henkilölle."
+    )
+    val fetched = Await.result(fetchKoulusivistyskieletForReal(missing), 30.seconds)
+    logger.info(
+      s"Koski palautti ${fetched.keys.size} henkilön koulusivistyskielen, tallennetaan välimuistiin."
+    )
```

Comment:
Voiskohan tän awaitin korvata sillä, että palauttaisi Futuren ja mappaisi sitä sitten eteenpäin kutsujan päässä, jossa jo valmiiksi pelataan futuureilla?

Reply:
Jostain syystä tuota kirjoittaessa jäin tosiaan arpomaan missä kohtaa jäisi odottamaan, mutta kun tuon muutoksen teki, kutsujan pään koodikin siistiytyi samalla ajatuksella ehkä taas hiukan:

https://github.com/Opetushallitus/hakurekisteri/pull/500/commits/5192f9a9a85d06f725cbf09d7dce45392ad98234